### PR TITLE
New DynamicData API [20168]

### PR DESCRIPTION
**Pull request by richiware**
## Contributor Checklist
- [ ] Commit messages follow the project guidelines. 
- [ ] The code follows the style guidelines of this project. 
- [ ] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- [ ] Any new/modified methods have been properly documented using Doxygen. 
- [ ] Changes are ABI compatible. 
- [ ] Changes are API compatible. 
- [ ] New feature has been added to the `versions.md` file (if applicable).
- [ ] New feature has been documented/Current behavior is correctly described in the documentation. 
    
- [ ] Applicable backports have been included in the description.

## Reviewer Checklist
- [ ] The PR has a milestone assigned.
- [ ] Check contributor checklist is correct.
- [ ] Check CI results: changes do not issue any warning.
- [ ] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Review by JLBuenoLopez-eProsima — CHANGES_REQUESTED** [review]

File: include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp [hunk 1]
```diff
@@ -0,0 +1,870 @@
...
+    virtual traits<DynamicType>::ref_type type() = 0;
+
+    /*!
+     * Retrieves the @ref MemberDescriptor associated to a member.
+     * @param[inout] value Non-nil @ref MemberDescriptor reference where the information is copied.
+     * @param[in] id Identifier of the member to be retrieved.
+     */
+    virtual ReturnCode_t get_descriptor(
+            traits<MemberDescriptor>::ref_type value,
+            MemberId id) = 0;
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp:48]: As you say, this was discussed and the conclusion was not include `set_descriptor`.

File: include/fastdds/dds/xtypes/dynamic_types/DynamicDataFactory.hpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+     * @return @ref DynamicDataFactory reference.
+     */
+    static traits<DynamicDataFactory>::ref_type get_instance();
+
+    /*!
+     * Resets the singleton reference.
+     * @return @ref ReturnCode_t
+     * @retval RETCODE_OK is always returned.
+     */
+    static ReturnCode_t delete_instance();
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicDataFactory.hpp:46]: Currently not known. When decide, we include it.

File: include/fastdds/dds/xtypes/dynamic_types/DynamicDataFactory.hpp [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+    /**
+     * Creates a new @ref DynamicData reference based on the given @ref DynamicType reference.
+     * @param[in] type @ref DynamicType reference associated.
+     * @return new @ref DynamicData reference
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicDataFactory.hpp]: Done in ffb72f347

File: include/fastdds/dds/xtypes/dynamic_types/DynamicDataFactory.hpp [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+     */
+    virtual traits<DynamicData>::ref_type create_data(
+            traits<DynamicType>::ref_type type) = 0;
+
+    /**
+     * Resets the internal reference if it is cached.
+     * @param[in] type @ref DynamicData reference whose internal cached reference to reset.
+     * @return standard ReturnCode_t
+     * @retval RETCODE_OK is always returned.
+     * @retval RETCODE_BAD_PARAMETER if reference is nil.
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicDataFactory.hpp]: Done in ffb72f347

File: src/cpp/fastdds/xtypes/dynamic_types/DynamicDataImpl.hpp [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+namespace fastdds {
+namespace dds {
+
+class DynamicDataImpl : public traits<DynamicData>::base_type
+{
+    traits<DynamicTypeImpl>::ref_type type_;
+
+    std::map<MemberId, std::shared_ptr<void>> value_;
+
+    std::vector<MemberId> loaned_values_;
```
[the remark this answers is not in the document]
Comment (richiware) [reply src/cpp/fastdds/xtypes/dynamic_types/DynamicDataImpl.hpp]: Done in ffb72f347

File: src/cpp/fastdds/xtypes/dynamic_types/DynamicDataImpl.hpp [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+    static size_t get_key_max_cdr_serialized_size(
+            traits<DynamicType>::ref_type type,
+            size_t current_alignment = 0);
+
+    static size_t get_max_cdr_serialized_size(
+            traits<DynamicType>::ref_type type,
+            size_t current_alignment = 0);
+
+    size_t get_cdr_serialized_size(
+            size_t current_alignment = 0) noexcept;
```
[the remark this answers is not in the document]
Comment (richiware) [reply src/cpp/fastdds/xtypes/dynamic_types/DynamicDataImpl.hpp]: Done in ffb72f347

File: src/cpp/fastdds/xtypes/dynamic_types/DynamicDataImpl.hpp [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+            MemberId id) noexcept;
+
+    void set_value(
+            const ObjectName& value,
+            MemberId id) noexcept;
+
+    void set_default_value(
+            MemberId id) noexcept;
+
+    ReturnCode_t clear_data() noexcept;
```
[the remark this answers is not in the document]
Comment (richiware) [reply src/cpp/fastdds/xtypes/dynamic_types/DynamicDataImpl.hpp]: Done in ffb72f347

File: include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp [hunk 8]
```diff
@@ -0,0 +1,870 @@
...
+
+class RTPS_DllAPI DynamicData : public std::enable_shared_from_this<DynamicData>
+{
+public:
+
+    using _ref_type = typename traits<DynamicData>::ref_type;
+
+    /*!
+     * Retrieve the @ref DynamicType reference associated to this @ref DynamicData
+     * @return @ref DynamicType reference
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp]: Done in ffb72f347

File: include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp [hunk 9]
```diff
@@ -0,0 +1,870 @@
...
+     * @param[in] id identifier of the object to retrieve
+     * @return @ref DynamicDataImpl reference loaned or \b nil on outstanding loaned data
+     */
+    virtual traits<DynamicData>::ref_type loan_value(
+            MemberId id) = 0;
+
+    /*!
+     * Returns a loaned retrieved using @ref DynamicData::loan_value.
+     * @param[in] value @ref DynamicData reference previously loaned
+     */
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp:159]: Done in ffb72f347

File: include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp [hunk 10]
```diff
@@ -0,0 +1,870 @@
...
+     * @return @ref ReturnCode_t
+     */
+    virtual ReturnCode_t clear_value(
+            MemberId id) = 0;
+
+    /*!
+     * \b Loans a @ref DynamicData reference within the sample
+     * @remarks This loan shall be returned by the @ref DynamicData::return_loaned_value operation
+     * @param[in] id identifier of the object to retrieve
+     * @return @ref DynamicDataImpl reference loaned or \b nil on outstanding loaned data
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp]: Done in ffb72f347

File: include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp [hunk 11]
```diff
@@ -0,0 +1,870 @@
...
+     * @li If the object is of a union type, return the number of members in the object. This number will be two if the
+     *     discriminator value selects a member and one otherwise.
+     * @li if the object is of a primitive or enumerated type, it is atomic: return one.
+     * @li if the object is of an alias type, return the value appropriate for the alias base type.
+     * @return count as defined above
+     */
+    virtual uint32_t get_item_count() = 0;
+
+    /*!
+     * Clear all memory associated to the object.
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp]: Done in ffb72f347

File: include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp [hunk 12]
```diff
@@ -0,0 +1,870 @@
...
+    virtual MemberId get_member_id_by_name(
+            const ObjectName& name) = 0;
+
+    /*!
+     * Queries @ref MemberId by index
+     * @param[in] index Index.
+     * @return MemberId or MEMBER_ID_INVALID on failure
+     */
+    virtual MemberId get_member_id_at_index(
+            uint32_t index) = 0;
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp]: Done in ffb72f347

File: include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp [hunk 13]
```diff
@@ -0,0 +1,870 @@
...
+            traits<DynamicData>::ref_type value) = 0;
+
+    /*!
+     * Retrieves a sequence of \b int32 values associated to an identifier.
+     * @param[inout] value \b Sequence of \b int32 to populate.
+     * @param[in] id Identifier of the member to query.
+     * @return @ref ReturnCode_t
+     */
+    virtual ReturnCode_t get_int32_values(
+            Int32Seq& value,
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp:612]: Done in ffb72f347

File: include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp [hunk 14]
```diff
@@ -0,0 +1,870 @@
...
+            bool value) = 0;
+
+    /*!
+     * Retrieves an \b string value associated to an identifier.
+     * @param[inout] value \b string to populate.
+     * @param[in] id Identifier of the member to query.
+     * @return @ref ReturnCode_t
+     */
+    virtual ReturnCode_t get_string_value(
+            std::string& value,
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp:540]: Done in ffb72f347

File: include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp [hunk 15]
```diff
@@ -0,0 +1,870 @@
...
+     * \b Loans a @ref DynamicData reference within the sample
+     * @remarks This loan shall be returned by the @ref DynamicData::return_loaned_value operation
+     * @param[in] id identifier of the object to retrieve
+     * @return @ref DynamicDataImpl reference loaned or \b nil on outstanding loaned data
+     */
+    virtual traits<DynamicData>::ref_type loan_value(
+            MemberId id) = 0;
+
+    /*!
+     * Returns a loaned retrieved using @ref DynamicData::loan_value.
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp]: Done in ffb72f347

File: include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp]: Done in ffb72f347

File: include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp [hunk 16]
```diff
@@ -0,0 +1,870 @@
...
+     */
+    virtual ReturnCode_t set_float128_value(
+            MemberId id,
+            long double value) = 0;
+
+    /*!
+     * Retrieves an \b char8 value associated to an identifier.
+     * @param[inout] value \b char8 to populate.
+     * @param[in] id Identifier of the member to query.
+     * @return standard DDS @ref ReturnCode_t
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp]: Done in ffb72f347

File: include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp [hunk 17]
```diff
@@ -0,0 +1,870 @@
...
+     *     discriminator value selects a member and one otherwise.
+     * @li if the object is of a primitive or enumerated type, it is atomic: return one.
+     * @li if the object is of an alias type, return the value appropriate for the alias base type.
+     * @return count as defined above
+     */
+    virtual uint32_t get_item_count() = 0;
+
+    /*!
+     * Clear all memory associated to the object.
+     * @return @ref ReturnCode_t
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp:114]: Done in ffb72f347

File: include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp [hunk 18]
```diff
@@ -0,0 +1,870 @@
...
+
+    /*!
+     * Provides the @b item @b count of the data and depends on the type of object:
+     * @li If the object is of a collection type, returns the number of elements currently in the collection.
+     *     In the case of an array type, this value will always be equal to the product of the bounds of all array
+     *     dimensions.
+     * @li If the object is of a bitmask type, return the number of named flags that are currently set in the bitmask.
+     * @li If the object is of a structure or annotation type, return the number of members in the object.
+     *     This value may be different than the number of members in the corresponding @ref DynamicType (some optional
+     *     members may be omitted.
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp]: Done in ffb72f347

File: include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp [hunk 19]
```diff
@@ -0,0 +1,870 @@
...
+     * Retrieve the @ref DynamicType reference associated to this @ref DynamicData
+     * @return @ref DynamicType reference
+     */
+    virtual traits<DynamicType>::ref_type type() = 0;
+
+    /*!
+     * Retrieves the @ref MemberDescriptor associated to a member.
+     * @param[inout] value Non-nil @ref MemberDescriptor reference where the information is copied.
+     * @param[in] id Identifier of the member to be retrieved.
+     */
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp:49]: Done in ffb72f347

File: include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp [hunk 20]
```diff
@@ -0,0 +1,1041 @@
...
+    virtual traits<DynamicData>::ref_type clone() = 0;
+
+    /*!
+     * Retrieves an \b int32 value associated to an identifier.
+     * @param[inout] value \b int32 to populate
+     * @param[in] Id identifier of the member to query.
+     * @return @ref ReturnCode_t
+     * @retval RETCODE_OK when the value was retrieved successfully.
+     * @retval RETCODE_BAD_PARAMETER when the @ref MemberId is invalid or the member type is not \b int32.
```
Comment (JLBuenoLopez-eProsima) [inline include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp:176]: According to section 7.5.2.11.9, there are implicit type promotions allowed. This suggestion applies to other similar comments in the methods below.

```suggestion
     * @retval RETCODE_BAD_PARAMETER when the @ref MemberId is invalid or the member type is not promotable to \b int32.
```

File: include/fastdds/dds/xtypes/dynamic_types/DynamicDataFactory.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicDataFactory.hpp:46]: Can we leave a TODO in order to remember and complete the documentation once it is clarified?

**Review by JLBuenoLopez-eProsima — COMMENTED** [review]
A couple of nitpicks

**Comment by JLBuenoLopez-eProsima** [comment]
@richiprosima please test this

File: include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp (hunk 20 above)
Comment (richiware) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp]: Done in 9be524d12

File: include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp [hunk 21]
```diff
@@ -0,0 +1,1041 @@
...
+     * @param[in] id identifier of the object to retrieve
+     * @return @ref DynamicData reference loaned or \b nil on outstanding loaned data
+     */
+    virtual traits<DynamicData>::ref_type loan_value(
+            MemberId id) = 0;
+
+    /*!
+     * Returns a loan retrieved using @ref DynamicData::loan_value.
+     * @param[in] value @ref DynamicData reference previously loaned
+     * @retval RETCODE_OK when the loaned was returned successfully.
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp]: Done in 9be524d12

File: include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp [hunk 22]
```diff
@@ -71,6 +79,13 @@ class RTPS_DllAPI DynamicData : public std::enable_shared_from_this<DynamicData>
 
     /*!
      * Queries @ref MemberId by index
+     * The query result depends on the type of the sample.
+     * Only next types support accessing by index.
+     * @li Aggregated type.
+     * @li Sequence type.
+     * @li Map type.
+     * @li Array type.
+     * @li Bitmask type.
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp]: Done in 9be524d12

File: include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp [hunk 23]
```diff
@@ -62,7 +66,11 @@ class RTPS_DllAPI DynamicData : public std::enable_shared_from_this<DynamicData>
             traits<DynamicData>::ref_type other) = 0;
 
     /*!
-     * Queries @ref MemberId by name
+     * Queries @ref MemberId by name.
+     * The query result depends on the type of the sample.
+     * Only next types support accessing by name.
+     * @li Aggregated type.
+     * @li Map type.
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp:73]: Done in 9be524d12

File: include/fastdds/dds/xtypes/dynamic_types/DynamicDataFactory.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicDataFactory.hpp]: Done in 9be524d12

File: include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp (hunk 20 above)
Comment (JLBuenoLopez-eProsima) [reply include/fastdds/dds/xtypes/dynamic_types/DynamicData.hpp]: This has been applied only to one of the methods. There are a lot of methods which this suggestion also applies

**Review by JLBuenoLopez-eProsima — APPROVED** [review]
LGTM

**Pull request merged**